src/equations/ideal_glm_mhd_multiion.jl

```diff
@@ -263,4 +263,171 @@ end
+        v3_avg = 0.5f0 * (v3_ll + v3_rr)
+        beta_avg = 0.5f0 * (beta_ll + beta_rr)
+        tau = 1 / (beta_ll + beta_rr)
+        p_mean = 0.5f0 * rho_avg / beta_avg
+        p_star = 0.5f0 * rho_ln / beta_ln
+        vel_norm_avg = 0.5f0 * (vel_norm_ll + vel_norm_rr)
+        vel_avg_norm = v1_avg^2 + v2_avg^2 + v3_avg^2
+        E_bar = p_star / (gammas[k] - 1) +
+                0.5f0 * rho_ln * (2 * vel_avg_norm - vel_norm_avg)
+
+        h11 = rho_ln
+        h12 = rho_ln * v1_avg
+        h13 = rho_ln * v2_avg
+        h14 = rho_ln * v3_avg
+        h15 = E_bar
+        d1 = -0.5f0 * λ *
+             (h11 * (w1_rr - w1_ll) +
+              h12 * (w2_rr - w2_ll) +
+              h13 * (w3_rr - w3_ll) +
+              h14 * (w4_rr - w4_ll) +
+              h15 * (w5_rr - w5_ll))
+
+        h21 = h12
+        h22 = rho_ln * v1_avg^2 + p_mean
+        h23 = h21 * v2_avg
+        h24 = h21 * v3_avg
+        h25 = (E_bar + p_mean) * v1_avg
+        d2 = -0.5f0 * λ *
+             (h21 * (w1_rr - w1_ll) +
+              h22 * (w2_rr - w2_ll) +
+              h23 * (w3_rr - w3_ll) +
+              h24 * (w4_rr - w4_ll) +
+              h25 * (w5_rr - w5_ll))
+
+        h31 = h13
+        h32 = h23
+        h33 = rho_ln * v2_avg^2 + p_mean
+        h34 = h31 * v3_avg
+        h35 = (E_bar + p_mean) * v2_avg
+        d3 = -0.5f0 * λ *
+             (h31 * (w1_rr - w1_ll) +
+              h32 * (w2_rr - w2_ll) +
+              h33 * (w3_rr - w3_ll) +
+              h34 * (w4_rr - w4_ll) +
+              h35 * (w5_rr - w5_ll))
+
+        h41 = h14
+        h42 = h24
+        h43 = h34
+        h44 = rho_ln * v3_avg^2 + p_mean
+        h45 = (E_bar + p_mean) * v3_avg
+        d4 = -0.5f0 * λ *
+             (h41 * (w1_rr - w1_ll) +
+              h42 * (w2_rr - w2_ll) +
+              h43 * (w3_rr - w3_ll) +
+              h44 * (w4_rr - w4_ll) +
+              h45 * (w5_rr - w5_ll))
+
+        h51 = h15
+        h52 = h25
+        h53 = h35
+        h54 = h45
+        h55 = ((p_star^2 / (gammas[k] - 1) + E_bar * E_bar) / rho_ln
+               +
+               vel_avg_norm * p_mean)
+        d5 = -0.5f0 * λ *
+             (h51 * (w1_rr - w1_ll) +
+              h52 * (w2_rr - w2_ll) +
+              h53 * (w3_rr - w3_ll) +
+              h54 * (w4_rr - w4_ll) +
+              h55 * (w5_rr - w5_ll))
+
+        beta_plus_ll += beta_ll
+        beta_plus_rr += beta_rr
+
+        set_component!(dissipation, k, d1, d2, d3, d4, d5, equations)
+    end
+
+    # Compute the dissipation related to the magnetic and divergence-cleaning fields
+    h_B_psi = 1 / (beta_plus_ll + beta_plus_rr)
+
+    # Dissipation due to the diagonal entries of the dissipation matrix H
+    dissipation[1] = -0.5f0 * λ * h_B_psi * (w_rr[1] - w_ll[1])
+    dissipation[2] = -0.5f0 * λ * h_B_psi * (w_rr[2] - w_ll[2])
+    dissipation[3] = -0.5f0 * λ * h_B_psi * (w_rr[3] - w_ll[3])
+    dissipation[end] = -0.5f0 * λ * h_B_psi * (w_rr[end] - w_ll[end])
```

Comment:
Should these use `-=` instead of just `=`? Otherwise, it seems that this overwrites the dissipation pieces that were computed and set in the loop above at the `set_component!` step.

Reply:
This does not overwrite the entries computed above. The reason is that the variable order is different than the one used by single-fluid MHD or multi-component MHD. The first three entries of the state vector `u[1:3]` are the magnetic field components. After that, we have blocks for the hydrodynamic quantities of each ion species. For instance, with two ion species we have the hydrodynamic quantities of species 1 in `u[4:8]` and the hydrodynamic quantities of species 2 in `u[9:13]`. Finally, the last entry `u[end]` is the divergence cleaning field. The `set_component!` function only writes hydrodynamic entries.

https://github.com/trixi-framework/Trixi.jl/blob/eaafa7306ed07bf1ace313b42a8c1b9b7f6c09e3/src/equations/ideal_glm_mhd_multiion.jl#L12-L22

Reply:
I just realized that this can be confusing because the ordering is different in the paper. The code's variable order was selected because it makes the index handling more convenient than the ordering of the paper. I'll improve the comments to make this clear.

Reply:
I improved the comments in 5aabeda40fc536aab7bae358ae3f5c5eccff57c3.